File: android/app/src/main/kotlin/com/simplephotos/data/album/AlbumSort.kt
/**
 * Album sort (#52) — the pure Date/Name ordering logic behind the album-detail
 * sort control, ported from web's `gallery/albumSort.ts`.
 *
 * Kept free of Android and Room types (it sorts via key selectors, not
 * `PhotoEntity`) so the comparator, the toggle logic and the persistence
 * serialization are all JVM-unit-testable — the same split `RenditionChoice.kt`
 * and `ClusterRename.kt` use. The Android glue (reading DataStore, driving the
 * grid) lives in `AlbumDetailViewModel`; the direction the sort is *applied* in
 * matches web exactly.
 */
package com.simplephotos.data.album

enum class AlbumSortField { DATE, NAME }
enum class AlbumSortDir { ASC, DESC }

data class AlbumSort(val field: AlbumSortField, val dir: AlbumSortDir)

/** The historical order before #52: capture date, newest first. Used as the
 *  control's visual default when the user has not chosen a sort. */
val DEFAULT_ALBUM_SORT = AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC)

/** The direction a field starts in when first switched to: dates newest-first,
 *  names A→Z — what a first tap on each is expected to do. */
fun defaultDirFor(field: AlbumSortField): AlbumSortDir =
    if (field == AlbumSortField.NAME) AlbumSortDir.ASC else AlbumSortDir.DESC

/**
 * Result of tapping a field button: toggle its direction if it is already the
 * active field, otherwise switch to it in its natural starting direction.
 * Mirrors web's `useAlbumSort.selectField`.
 */
fun nextSort(current: AlbumSort, field: AlbumSortField): AlbumSort =
    if (current.field == field) {
        current.copy(
            dir = if (current.dir == AlbumSortDir.ASC) AlbumSortDir.DESC else AlbumSortDir.ASC
        )
    } else {
        AlbumSort(field, defaultDirFor(field))
    }

/**
 * Total order over album items, tie-broken on a stable id so equal keys never
 * reorder between renders. `takenAt`/`name`/`id` are selectors so this needs no
 * Android type and can be unit-tested with plain data.
 *
 * A missing capture time (0) sorts as the oldest — last under desc, first under
 * asc — rather than landing somewhere arbitrary.
 */
fun <T> sortAlbumItems(
    items: List<T>,
    sort: AlbumSort,
    takenAt: (T) -> Long,
    name: (T) -> String,
    id: (T) -> String,
): List<T> {
    val comparator = Comparator<T> { a, b ->
        var c =
            if (sort.field == AlbumSortField.NAME) naturalCompare(name(a), name(b))
            else takenAt(a).compareTo(takenAt(b))
        if (c == 0) c = id(a).compareTo(id(b))
        if (sort.dir == AlbumSortDir.ASC) c else -c
    }
    return items.sortedWith(comparator)
}

/**
 * Natural, case-insensitive, numeric-aware string comparison so `IMG_2` precedes
 * `IMG_10` instead of sorting lexically. Mirrors web's
 * `Intl.Collator(numeric: true, sensitivity: "base")`: digit runs compare as
 * numbers (leading zeros ignored), everything else compares case-folded.
 */
internal fun naturalCompare(a: String, b: String): Int {
    val x = a.lowercase()
    val y = b.lowercase()
    var i = 0
    var j = 0
    while (i < x.length && j < y.length) {
        val cx = x[i]
        val cy = y[j]
        if (cx.isDigit() && cy.isDigit()) {
            var si = i
            while (si < x.length && x[si].isDigit()) si++
            var sj = j
            while (sj < y.length && y[sj].isDigit()) sj++
            // Compare numerically by stripping leading zeros, then by length,
            // then lexically — avoids overflow on very long digit runs.
            val numX = x.substring(i, si).trimStart('0')
            val numY = y.substring(j, sj).trimStart('0')
            if (numX.length != numY.length) return numX.length - numY.length
            val c = numX.compareTo(numY)
            if (c != 0) return c
            i = si
            j = sj
        } else {
            if (cx != cy) return cx.compareTo(cy)
            i++
            j++
        }
    }
    return (x.length - i) - (y.length - j)
}

// ── Persistence serialization (DataStore stores the string) ─────────────────

/** `"date:desc"` / `"name:asc"`. */
fun AlbumSort.serialize(): String =
    "${field.name.lowercase()}:${dir.name.lowercase()}"

/** Parse a stored value; a missing or malformed one reads as "no choice"
 *  (null), which keeps the album in its intrinsic order. */
fun parseAlbumSort(raw: String?): AlbumSort? {
    if (raw.isNullOrBlank()) return null
    val parts = raw.split(":")
    if (parts.size != 2) return null
    val field = when (parts[0]) {
        "date" -> AlbumSortField.DATE
        "name" -> AlbumSortField.NAME
        else -> return null
    }
    val dir = when (parts[1]) {
        "asc" -> AlbumSortDir.ASC
        "desc" -> AlbumSortDir.DESC
        else -> return null
    }
    return AlbumSort(field, dir)
}

File: web/src/gallery/albumSort.ts
/**
 * Album sort (#52) — a user-selectable ordering for the album-detail view.
 *
 * Ordering used to be hardcoded: `takenAt` desc everywhere, with the single
 * `sortBy: "addedAt"` special case for "Recently Added". This adds a Date/Name
 * sort control to the album header. The choice is per-album and applied
 * uniformly in `useAlbumPhotos.resolveAlbumPhotos` (never special-cased per
 * view), **after** burst collapse so a burst sorts by its representative frame.
 *
 * The sort is deliberately *optional*: when a user has made no choice, the
 * album keeps its intrinsic order (which is what preserves "Recently Added"'s
 * add-order). Only an explicit selection triggers a post-sort.
 */
import type { PhotoWithBurstCount } from "../utils/burstCollapse";

export type SortField = "date" | "name";
export type SortDir = "asc" | "desc";
export interface AlbumSort {
  field: SortField;
  dir: SortDir;
}

/** The historical ordering before #52: capture date, newest first. Used as the
 *  control's visual default when the user has not chosen a sort. */
export const DEFAULT_ALBUM_SORT: AlbumSort = { field: "date", dir: "desc" };

/** The direction a field starts in when you first switch to it: dates
 *  newest-first, names A→Z — what a first click on each is expected to do. */
export function defaultDirFor(field: SortField): SortDir {
  return field === "name" ? "asc" : "desc";
}

// `numeric: true` so "IMG_2" precedes "IMG_10" instead of sorting lexically;
// `sensitivity: "base"` makes it case- and accent-insensitive, matching how a
// file browser orders names.
const collator = new Intl.Collator(undefined, {
  numeric: true,
  sensitivity: "base",
});

/**
 * Total order over resolved album photos. Always breaks ties on `blobId` so the
 * result is deterministic — two photos sharing a capture time or filename never
 * swap places between renders, which is the stability `JustifiedGrid` needs to
 * keep the scroll position from jumping.
 */
export function compareAlbumPhotos(
  a: PhotoWithBurstCount,
  b: PhotoWithBurstCount,
  sort: AlbumSort
): number {
  let cmp: number;
  if (sort.field === "name") {
    cmp = collator.compare(a.filename ?? "", b.filename ?? "");
  } else {
    // A missing capture time (0) sorts as the oldest — last under desc, first
    // under asc — rather than throwing the row to a random position.
    cmp = (a.takenAt ?? 0) - (b.takenAt ?? 0);
  }
  if (cmp === 0) {
    cmp = a.blobId < b.blobId ? -1 : a.blobId > b.blobId ? 1 : 0;
  }
  return sort.dir === "asc" ? cmp : -cmp;
}

/** Return a new, sorted array — never mutates the input list. */
export function sortAlbumPhotos<T extends PhotoWithBurstCount>(
  photos: T[],
  sort: AlbumSort
): T[] {
  return [...photos].sort((a, b) => compareAlbumPhotos(a, b, sort));
}

// ── Persistence (per album, localStorage) ──────────────────────────────────
// Keyed by album id, matching the per-view `useScrollMemory` precedent. A
// regular album's manifest id and a smart album's synthetic id are both stable,
// so the choice survives navigation and restarts.

const keyFor = (albumId: string) => `albumSort:${albumId}`;

/** The user's chosen sort for this album, or `null` if they have not set one
 *  (in which case the album keeps its intrinsic order). Tolerates absent or
 *  corrupt storage — a bad value reads as "no choice", never throws. */
export function readAlbumSort(albumId: string): AlbumSort | null {
  try {
    const raw = localStorage.getItem(keyFor(albumId));
    if (!raw) return null;
    const parsed: unknown = JSON.parse(raw);
    return isAlbumSort(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

export function writeAlbumSort(albumId: string, sort: AlbumSort): void {
  try {
    localStorage.setItem(keyFor(albumId), JSON.stringify(sort));
  } catch {
    // Storage full or blocked (private mode): the choice still applies this
    // session via component state; only its persistence is lost.
  }
}

export function isAlbumSort(v: unknown): v is AlbumSort {
  if (!v || typeof v !== "object") return false;
  const s = v as Partial<AlbumSort>;
  return (
    (s.field === "date" || s.field === "name") &&
    (s.dir === "asc" || s.dir === "desc")
  );
}
